src/register/pmpaddrx.rs

```diff
@@ -13,15 +13,15 @@ pub type Addr = u64;
 
 impl PmpAddr {
     #[inline]
-    pub fn decode(&self, mode: Mode) -> (Addr, Option<NonZeroSize>) {
```

Comment:
This type is getting fairly complex now, and it also has meaningless combinations, like (None, Some) which can never be returned. WDYT about adding a new type to represent the return value for this?

Reply:
I am not necessarily opposed.  Were you thinking something like an `enum`, I am not sure if a struct would allow us to remove the meaningless combinations?  My worry is that it would complicate using the library.  To extract values from the enum you need to `if let` or `match` on the enum, adding a fair bit of boilerplate code.
The only user of this interface right now(?) is humility in the [pmp decode functions](https://github.com/rivosinc/humility/blob/4b863cc56dfa817f5c3e9fb1cbdc0b11d226f5bd/cmd/pmp/src/lib.rs#L124) (just for some context of how this is used).

Reply:
I think making an enum would be a good move here. Maybe change the `Mode` enum to something like this?

```rust

enum Mode {
    OFF,
    TOR(Addr),
    NAPOT(Addr, NonZeroSize),
    NA4(Addr, NonZeroSize)
}
```

This would require a bigger refactor of the pmp* modules, but I think the API would be a bit cleaner this way. Instead of having to provide a `mode` argument, the function can simply check on its own and construct the correct enum type. It should also be fairly trivial to switch Humility to use this.